python/ray/tune/schedulers/pb2_utils.py

```diff
@@ -0,0 +1,177 @@
+        timekernel = (1 - self.epsilon)**(0.5 * dists)
+
+        X = X[:, 1:]
+        X2 = X2[:, 1:]
+
+        RBF = self.variance * np.exp(
+            -np.square(euclidean_distances(X, X2)) / self.lengthscale)
+
+        return RBF * timekernel
+
+    def Kdiag(self, X):
+        return self.variance * np.ones(X.shape[0])
+
+    def update_gradients_full(self, dL_dK, X, X2):
+        if X2 is None:
+            X2 = np.copy(X)
+        T1 = X[:, 0].reshape(-1, 1)
+        T2 = X2[:, 0].reshape(-1, 1)
+
+        X = X[:, 1:]
+        X2 = X2[:, 1:]
+        dist2 = np.square(euclidean_distances(X, X2)) / self.lengthscale
+
+        dvar = np.exp(-np.square(
+            (euclidean_distances(X, X2)) / self.lengthscale))
+        dl = -(2 * euclidean_distances(X, X2)**2 * self.variance *
+               np.exp(-dist2)) * self.lengthscale**(-2)
+        n = pairwise_distances(T1, T2, 'cityblock') / 2
+        deps = -n * (1 - self.epsilon)**(n - 1)
+
+        self.variance.gradient = np.sum(dvar * dL_dK)
+        self.lengthscale.gradient = np.sum(dl * dL_dK)
+        self.epsilon.gradient = np.sum(deps * dL_dK)
+
+
+def normalize(data, wrt):
+    """ Normalize data to be in range (0,1), with respect to (wrt) boundaries, 
+        which can be specified.
+    """
+    return (data - np.min(wrt, axis=0)) / (
+        np.max(wrt, axis=0) - np.min(wrt, axis=0))
+
+
+def standardize(data):
+    """ Standardize to be Gaussian N(0,1). Clip final values.
+    """
+    data = (data - np.mean(data, axis=0)) / (np.std(data, axis=0) + 1e-8)
+    return np.clip(data, -2, 2)
+
+def UCB(m, m1, x, fixed, kappa=0.5):
+    """ UCB acquisition function. Interesting points to note:
+        1) We concat with the fixed points, because we are not optimizing wrt
+           these. This is the Reward and Time, which we can't change. We want
+           to find the best hyperparameters *given* the reward and time.
+        2) We use m to get the mean and m1 to get the variance. If we already
+           have trials running, then m1 contains this information. This reduces
+           the variance at points currently running, even if we don't have their
+           label. 
+           Ref: https://jmlr.org/papers/volume15/desautels14a/desautels14a.pdf
+
+    """
+
+    c1 = 0.2
+    c2 = 0.4
+    beta_t = c1 * np.log(c2 * m.X.shape[0])
+    kappa = np.sqrt(beta_t)
+
+    xtest = np.concatenate((fixed.reshape(-1, 1), np.array(x).reshape(-1,
+                                                                      1))).T
+
+    preds = m.predict(xtest)
+    mean = preds[0][0][0]
+
+    preds = m1.predict(xtest)
+    var = preds[1][0][0]
+    return mean + kappa * var
+
+
+def optimize_acq(func, m, m1, fixed, num_f):
+    """ Optimize acquisition function.
+    """
+    
+    opts = {'maxiter': 200, 'maxfun': 200, 'disp': False}
+
+    T = 10
+    best_value = -999
+    best_theta = m1.X[0, :]
+
+    bounds = [(0, 1) for _ in range(m.X.shape[1] - num_f)]
```

Comment:
completely untested:
```suggestion
    bounds = [(0, 1), ]*(m.X.shape[1] - num_f)
```

Reply:
Why `[(0,1), ]` instead of just `[(0,1)]`?

Reply:
no particular reason, I just tend to add the comma for collections that have a single element out of personal preference. I do this because `('hello')` becomes `'hello'`, not a tuple as "expected"